s2spy/traintest.py

```diff
@@ -1,105 +1,115 @@
+    multiple input arrays at once. They need to have matching coordinates along
+    the given dimension.
 
-    group_labels = np.unique(data[key])
-    if group_labels.size <= 1:
-        raise ValueError(f"Input data must have more than 1 unique {key} to split data.")
-
-    splits = splitter.split(group_labels)
-
-    # label every row in the dataframe
-    for i, (train_indices, test_indices) in enumerate(splits):
-        # create column split_# filled with "skip" in dataframe
-        col_name = f"split_{i}"
-        data[col_name] = ""
-
-        # create dictionary with key and train/test labels
-        indices = np.empty(group_labels.size, dtype="<U6")
-        indices[:] = "skip"
-        indices[train_indices] = "train"
-        indices[test_indices] = "test"
-        train_test_dict = dict(zip(group_labels, indices))
-
-        # map key row values with train/test
-        data[col_name] = data[key].map(train_test_dict)
+    For an overview of the sklearn Splitter Classes see:
+    https://scikit-learn.org/stable/modules/classes.html#module-sklearn.model_selection
 
-    return data
+    Args:
+        splitter (SplitterClass): Initialized splitter class, much have a
+            `fit(X)` method which splits up `X` into multiple folds of
+            train/test data.
+    """
+    def __init__(self, splitter: Type[BaseCrossValidator]) -> None:
+        self.splitter = splitter
+
+    def split(
+        self,
+        *x_args: xr.DataArray,
+        y: Optional[xr.DataArray] = None,
+        dim: str = "anchor_year"
+    ) -> XMaybeY:
+        """Iterate over splits.
+
+        Args:
+            x_args: one or multiple xr.DataArray's that share the same
+                coordinate along the given dimension
+            y: (optional) xr.DataArray that shares the same coordinate along the
+                given dimension
+            dim: name of the dimension along which to split the data.
+
+        Returns:
+            Iterator over the splits
+        """
+        # Check that all inputs share the same dim coordinate
+        coords = []
+        x: xr.DataArray  # Initialize x to set scope outside loop
+        for x in x_args:
+            try:
+                coords.append(x[dim])
+            except KeyError as err:
+                raise CoordinateMismatch(
+                    f"Not all input data arrays have the {dim} dimension."
+                ) from err
+
+        if not _all_equal(coords):
+            raise CoordinateMismatch(
+                f"Input arrays are not equal along {dim} dimension."
+                )
+
+        if y is not None and not np.array_equal(y[dim], x[dim]):
+            raise CoordinateMismatch(
+                f"Input arrays are not equal along {dim} dimension."
+            )
+
+        if x[dim].size <=1:
+            raise ValueError(
+                f"Cannot split: need at least 2 values along dimension {dim}"
+            )
```

Comment:
Not so sure if this is needed. The splitters in sklearn will take care of this. For instance, if we only have one value and call k-folds method (e.g. n_splits=3) to split it, it will complain:

```py
# create data
n = 8
time_index = pd.date_range("20151020", periods=n, freq="60d")
time_coord = {"time": time_index}
x1 = xr.DataArray(np.random.randn(n), coords=time_coord, name="precursor1")
# resample
calendar = s2spy.time.AdventCalendar(anchor=(10, 15), freq="180d")
calendar.map_to_data(x1) 
x1 = s2spy.time.resample(calendar, x1)
# cross validate
kfold = KFold(n_splits=3)
cv = s2spy.traintest.TrainTestSplit(kfold)
for x1_train, x1_test in cv.split(x1):
    print("Train:", x1_train.anchor_year.values)
    print("Test:", x1_test.anchor_year.values)

>>> ValueError: Cannot have number of splits n_splits=3 greater than the number of samples: n_samples=1.
```

I would suggest to remove this and let the splitters take care of it, as the error message is more informative with different splitter.

Reply:
In the tests (`test_kfold_too_short`), this produces a much more obscure error message: 
```
--> 325             raise TypeError(
    326                 "Singleton array %r cannot be considered a valid collection." % x
    327             )
```

Hence I think it is worth adding our own, custom error message. However, since sklearn apparently raises different types of errors (ValueError or TypeError), rather than catching it and re-raising, we might be better of with just our own small sanity check. I adapted the message slightly.